### ISSUE TITLE
Add support for AWS Secrets Manager as an external key value storage

### ISSUE THREAD

**Issue by jonathanfoster** [TykTechnologies/tyk#6563]
**Is your feature request related to a problem? Please describe.**

I'm an enterprise Tyk Cloud customer and I have a ton of APIs that require auth tokens for upstream API authorization. I use AWS Secrets Manager for external secrets, but this isn't natively supported by Tyk. Instead I have to run External Secrets Operator to inject secrets as environment variables, which requires me to store all secret key-value pairs in one external secret (e.g., `TYK_GW_SECRETS=key:value,key2:value2`). There's no way for me to allow individual API owners to manage their own secrets.

**Describe the solution you'd like**

A solution would be to natively support AWS Secrets Manager as an external key-value store.

**Describe alternatives you've considered**

The current workaround is to use External Secrets Operator to map secrets from AWS to environment variables in Tyk. This has the limitation of only supporting one secret with all key-value pairs.

Another option I've consider is installing Vault and have it integrate with AWS. This would require an enterprise license for Vault though.

**Comment by jonathanfoster** [comment]
I started looking into implementing this myself and found I can inject individual secrets as environment variables named `TYK_SECRET_${key}`, there's no need to combine all secrets into key-value pairs in `TYK_GW_SECRETS`. This certainly makes secrets more manageable with ESO, but it would still be nice if Secrets Manager was natively supported.

The `TYK_SECRET_${key}` feature wasn't 100% clear after initially reading the [KV store docs](https://tyk.io/docs/tyk-configuration-reference/kv-store/#environment-variables) so maybe there's an opportunity to clarify. Actually, I would say that's true for the KV store docs in general. It's been tough to wrap my head around what notation is required for what location and whether the reference supports a path or uses the hardcoded `tyk-apis` path.

**Comment by andyo-tyk** [comment]
Hi @jonathanfoster,

Thanks for raising this ticket. We are reviewing the enhancements we made to KV storage earlier this year and looking towards the future possibilities (which include improving support for other storage options).

In terms of the documentation, it's a tricky subject to cover concisely at the moment, due to some details of the implementation, but we have tried to be as clear as possible what notation is supported in different scenarios.
If you've got some specific questions or suggestions we could look to clarify further.

Thanks for your support.

**Comment by jonathanviber** [comment]
@jonathanfoster , i'm also using external secrets but I want to rotate secrets so using ENV vars means I have to restart the pod. Any alternative solutions?

**Comment by jonathanfoster** [comment]
@jonathanviber You can use https://github.com/stakater/Reloader to automatically restart workloads and load the new config values.

**Comment by jonathanviber** [comment]
@jonathanfoster , we tried this using a CRON job and it worked but we rotate certain keys very frequently and we rely very heavily on our API Gateway so we didn't really like it and felt like there should be an easier way. We had the same issue with KrakenD so we wrote a Go plugin which retrieved the secret from a local file once every few minutes and then kept the value in cache. We may have to do the same here. I was hoping there was an easier way. It would be great if the tyk conf "secrets" supported getting a secret from file with a refresh frequency.

**Comment by jonathanfoster** [comment]
Yeah, reloader can be disruptive if secrets are volatile. Mounting secrets as a volume is an even better idea though since they can’t be easily exposed in resource manifests like env vars. You would probably just need to hot reload the gateway or API for the new values to take effect.

**Comment by jonathanviber** [comment]
Our plugin for KrakenD didn't require a reload since the plugin worked pretty independently. Hopefully we will get the same working here.
